### ISSUE TITLE
Why weight restoring is not the default option in EarlyStopping?

### ISSUE THREAD

**Issue by karkirowle** [keras-team/keras#11371]
I was looking over the Keras documentation, when I noticed that the default for keras.callbacks.EarlyStopping is restore_best_weights=False.

In the documentation, the described behaviour is:
restore_best_weights: whether to restore model weights from the epoch with the best value of the monitored quantity. If False, the model weights obtained at the last step of training are used.

I find this a bit confusing because I think the principle of early stopping is to obtain the best results on the validation set by stopping the optimisation when the validation loss increases, which is the case without setting the patience parameter.

However, in early stopping it makes sense to set the patience parameter because you might get spurious overfittings due to momentary local optima. In that case, you would definitely want to restore the previous weights, because you might severely overfit after the end of your patience period.
Furthermore, I could not think of any reason why you would not want restore best weights in Early Stopping. The only case I can think of when you want to stop your training when there is no significant improvement on your loss, in order to reduce training time, but this is not the purpose of early stopping.
  
I'm wondering if there is any use case at all to make the False the default behaviour and if not, would not it make sense to change this?

**Comment by jjerphan** [comment]
If you want to restore the weights that are giving the best performance, you have to keep tracks on them and thus have to store them. This can be costly as you have to keep another entire model in memory and can make fitting the model slower as well.

I think this is why the default behavior is to stop when the monitored metric isn't improving anymore — hence the default small `patience` to try to keep a good model.

**Comment by karkirowle** [comment]
I see the reason now, but I still don't think that this is a good default option. I do see that you have to keep another model in memory, and this can be costly, but the model could have moved away a lot from the local optima in the patience period, meaning the model needs to be retrained from the beginning or from the last saved checkpoint, which is more costly. I think the default behaviour should be emphasised more in the documentation also when explaining the restore_best_weights, not only in the default function call, because it is very easy to look over this, due to the scrollbar.

**Comment by jjerphan** [comment]
Maybe the documentation just need to be a bit clearer on this point.

**Comment by gat1s** [comment]
I thought I'd have to do model checkpointing and early stopping - both - to be able to stop training early and save the model with the best performance. 

Until I realised that you can just do early stopping with restore_best_weights changed to True.
